### PR TITLE
Feature/dashboard duplication

### PR DESCRIPTION
**Pull request by natesh-chidhambaram**
[MIP-324](https://datakrew.atlassian.net/browse/MIP-324)
**Feature:** Dashboard duplication.
User can duplicate existing dashboard by providing the dashboard_id and a name. The given name must not be present under the organisation.

### CODE REVIEW

File: apps/acqdat_api/lib/acqdat_api/dashboard_management/dashboard.ex [hunk 1]
```diff
@@ -51,6 +53,42 @@ defmodule AcqdatApi.DashboardManagement.Dashboard do
...
+              {:error, Reason} ->
+                {:error, Reason}
+            end
+          end)
+
+          DashboardModel.get_with_panels(new_dashboard.id)
+
+        {:error, Reason} ->
+          {:error, Reason}
+      end
```
Comment (niteshcode4u) [inline apps/acqdat_api/lib/acqdat_api/dashboard_management/dashboard.ex:88]: Let's use with clause and remove nested case structure

File: apps/acqdat_api/lib/acqdat_api/dashboard_management/dashboard.ex [hunk 2]
```diff
@@ -51,6 +53,42 @@ defmodule AcqdatApi.DashboardManagement.Dashboard do
...
+                  create_widget_instance_in_panel(panel_details.widget_instances, new_panel)
+                end
+
+                if panel_details.children != [] do
+                  subpanel_attrs = create_subpanel_attributes(panel_details.children, new_panel)
+                  Repo.insert_all(Panel, subpanel_attrs)
+                end
+
+              {:error, Reason} ->
+                {:error, Reason}
```
Comment (niteshcode4u) [inline apps/acqdat_api/lib/acqdat_api/dashboard_management/dashboard.ex:80]: typo: `Reason` variable should be named with lower case `reason`

File: apps/acqdat_api/lib/acqdat_api/dashboard_management/dashboard.ex [hunk 3]
```diff
@@ -51,6 +53,42 @@ defmodule AcqdatApi.DashboardManagement.Dashboard do
...
+            end
+          end)
+
+          DashboardModel.get_with_panels(new_dashboard.id)
+
+        {:error, Reason} ->
+          {:error, Reason}
```
Comment (niteshcode4u) [inline apps/acqdat_api/lib/acqdat_api/dashboard_management/dashboard.ex:87]: typo: `Reason` variable should be named with lower case `reason`

File: apps/acqdat_api/lib/acqdat_api_web/controllers/dashboard_management/dashboard_controller.ex [hunk 4]
```diff
@@ -102,6 +105,47 @@ defmodule AcqdatApiWeb.DashboardManagement.DashboardController do
     end
   end
 
+  def duplicate_dashboard(conn, params) do
+    case conn.status do
+      nil ->
+        params =
+          params
+          |> Map.put_new("creator_id", String.to_integer(Guardian.Plug.current_resource(conn)))
```
Comment (niteshcode4u) [inline apps/acqdat_api/lib/acqdat_api_web/controllers/dashboard_management/dashboard_controller.ex:113]: please remove single piping

File: apps/acqdat_api/lib/acqdat_api_web/controllers/dashboard_management/dashboard_controller.ex [hunk 5]
```diff
@@ -102,6 +105,47 @@ defmodule AcqdatApiWeb.DashboardManagement.DashboardController do
...
+                true -> message.error
+              end
+
+            send_error(conn, 400, response)
+        end
+
+      404 ->
+        conn
+        |> send_error(404, DashboardErrorHelper.error_message(:resource_not_found))
+
```
Comment (niteshcode4u) [inline apps/acqdat_api/lib/acqdat_api_web/controllers/dashboard_management/dashboard_controller.ex:142]: please remove single piping

File: apps/acqdat_api/lib/acqdat_api_web/controllers/dashboard_management/dashboard_controller.ex [hunk 6]
```diff
@@ -102,6 +105,47 @@ defmodule AcqdatApiWeb.DashboardManagement.DashboardController do
...
+      401 ->
+        conn
+        |> send_error(401, DashboardErrorHelper.error_message(:unauthorized))
```
Comment (niteshcode4u) [inline apps/acqdat_api/lib/acqdat_api_web/controllers/dashboard_management/dashboard_controller.ex:145]: please remove single piping

File: apps/acqdat_api/test/acqdat_api_web/controllers/dashboard-management/dashboard_controller_test.exs [hunk 7]
```diff
@@ -584,4 +585,102 @@ defmodule AcqdatApiWeb.DashboardManagement.DashboardControllerTest do
...
+      assert Map.has_key?(response, "name")
+      assert Map.has_key?(response, "id")
+    end
+
+    test "fails if authorization header not found", %{conn: conn, dashboard: dashboard} do
+      bad_access_token = "qwerty1234567uiop"
+
+      conn =
+        conn
+        |> put_req_header("authorization", "Bearer #{bad_access_token}")
```
Comment (niteshcode4u) [inline apps/acqdat_api/test/acqdat_api_web/controllers/dashboard-management/dashboard_controller_test.exs:620]: please remove single piping

File: apps/acqdat_api/test/acqdat_api_web/controllers/dashboard-management/dashboard_controller_test.exs [hunk 8]
```diff
@@ -584,4 +585,102 @@ defmodule AcqdatApiWeb.DashboardManagement.DashboardControllerTest do
...
+      data = %{dashboard_id: dashboard.id}
+
+      conn =
+        post(
+          conn,
+          Routes.dashboard_path(conn, :duplicate_dashboard, dashboard.org_id, dashboard.id),
+          data
+        )
+
+      result = conn |> json_response(403)
```
Comment (niteshcode4u) [inline apps/acqdat_api/test/acqdat_api_web/controllers/dashboard-management/dashboard_controller_test.exs:631]: please remove single piping

File: apps/acqdat_api/test/acqdat_api_web/controllers/dashboard-management/dashboard_controller_test.exs [hunk 9]
```diff
@@ -584,4 +585,102 @@ defmodule AcqdatApiWeb.DashboardManagement.DashboardControllerTest do
...
+      data = %{dashboard_id: dashboard.id}
+
+      conn =
+        post(
+          conn,
+          Routes.dashboard_path(conn, :duplicate_dashboard, dashboard.org_id, dashboard.id),
+          data
+        )
+
+      response = conn |> json_response(400)
```
Comment (niteshcode4u) [inline apps/acqdat_api/test/acqdat_api_web/controllers/dashboard-management/dashboard_controller_test.exs:651]: please remove single piping

File: apps/acqdat_api/test/acqdat_api_web/controllers/dashboard-management/dashboard_controller_test.exs [hunk 10]
```diff
@@ -584,4 +585,102 @@ defmodule AcqdatApiWeb.DashboardManagement.DashboardControllerTest do
...
+      }
+
+      conn =
+        post(
+          conn,
+          Routes.dashboard_path(conn, :duplicate_dashboard, panel.org_id, panel.dashboard_id),
+          data
+        )
+
+      response = conn |> json_response(200)
```
Comment (niteshcode4u) [inline apps/acqdat_api/test/acqdat_api_web/controllers/dashboard-management/dashboard_controller_test.exs:677]: please remove single piping

File: apps/acqdat_api/test/acqdat_api_web/controllers/dashboard-management/dashboard_controller_test.exs [hunk 11]
```diff
@@ -584,4 +585,102 @@ defmodule AcqdatApiWeb.DashboardManagement.DashboardControllerTest do
...
+      assert Map.has_key?(response, "name")
+      assert Map.has_key?(response, "id")
+      new_id = response["id"]
+      query = from(s in Panel, where: s.dashboard_id == ^new_id)
```
Comment (niteshcode4u) [inline apps/acqdat_api/test/acqdat_api_web/controllers/dashboard-management/dashboard_controller_test.exs:681]: `s in Panel` should be `p in Panel`

File: apps/acqdat_api/test/acqdat_api_web/controllers/dashboard-management/dashboard_controller_test.exs [hunk 12]
```diff
@@ -584,4 +585,102 @@ defmodule AcqdatApiWeb.DashboardManagement.DashboardControllerTest do
...
+      db_resp = Repo.all(query)
```
Comment (niteshcode4u) [inline apps/acqdat_api/test/acqdat_api_web/controllers/dashboard-management/dashboard_controller_test.exs:682]:
```suggestion
     db_resp = 
       (p in Panel)
       |> from(where: p.dashboard_id == ^new_id)
       |> Repo.all()
```

**Review by niteshcode4u — CHANGES_REQUESTED** [review]

File: apps/acqdat_api/lib/acqdat_api/dashboard_management/dashboard.ex [hunk 13]
```diff
@@ -111,4 +156,126 @@ defmodule AcqdatApi.DashboardManagement.Dashboard do
...
   end
+
+  defp duplicate_dashboard_params(dashboard, %{name: name, creator_id: creator_id}) do
+    %{
+      name: name,
+      description: dashboard.description,
+      org_id: dashboard.org_id,
+      avatar: dashboard.avatar,
+      settings: dashboard.settings || %{},
+      settings: (dashboard.settings && Map.from_struct(dashboard.settings)) || %{},
```
Comment (BandanaPandey) [inline apps/acqdat_api/lib/acqdat_api/dashboard_management/dashboard.ex:167]: @natesh-chidhambaram why settings key is duplicated here ?

File: apps/acqdat_api/lib/acqdat_api/dashboard_management/dashboard.ex [hunk 14]
```diff
@@ -111,4 +156,126 @@ defmodule AcqdatApi.DashboardManagement.Dashboard do
...
+           description: description,
+           org_id: org_id,
+           settings: settings,
+           widget_layouts: widget_layouts,
+           icon: icon,
+           filter_metadata: filter_metadata
+         },
+         dashboard_id
+       ) do
+    datetime = DateTime.truncate(DateTime.utc_now(), :second)
```
Comment (BandanaPandey) [inline apps/acqdat_api/lib/acqdat_api/dashboard_management/dashboard.ex:184]: @natesh-chidhambaram  where are you using this datetime var ?

File: apps/acqdat_api/lib/acqdat_api/dashboard_management/dashboard.ex [hunk 15]
```diff
@@ -111,4 +156,126 @@ defmodule AcqdatApi.DashboardManagement.Dashboard do
...
+           icon: icon,
+           description: description,
+           org_id: org_id,
+           settings: settings,
+           filter_metadata: filter_metadata,
+           widget_layouts: widget_layouts
+         },
+         panel
+       ) do
+    datetime = DateTime.truncate(DateTime.utc_now(), :second)
```
Comment (BandanaPandey) [inline apps/acqdat_api/lib/acqdat_api/dashboard_management/dashboard.ex:261]: same, where are we using this datetime ?

File: apps/acqdat_api/lib/acqdat_api_web/router.ex [hunk 16]
```diff
@@ -172,8 +172,11 @@ defmodule AcqdatApiWeb.Router do
           end
 
           post "/duplicates", PanelController, :duplicate_panels
+          post "/duplicate_dashboards", DashboardController, :duplicate_dashboard
         end
 
+        # post "/duplicate_dashboards", DashboardController, :duplicate_dashboard
```
Comment (BandanaPandey) [inline apps/acqdat_api/lib/acqdat_api_web/router.ex:178]: @natesh-chidhambaram  please remove commented code

File: apps/acqdat_api/lib/acqdat_api_web/router.ex [hunk 17]
```diff
@@ -172,8 +172,11 @@ defmodule AcqdatApiWeb.Router do
           end
 
           post "/duplicates", PanelController, :duplicate_panels
+          post "/duplicate_dashboards", DashboardController, :duplicate_dashboard
```
Comment (BandanaPandey) [inline apps/acqdat_api/lib/acqdat_api_web/router.ex:175]: @natesh-chidhambaram  also make it plural duplicate_dashboard should be duplicate_dashboards

File: apps/acqdat_api/lib/acqdat_api/dashboard_management/dashboard.ex (hunk 15 above)
Comment (natesh-chidhambaram) [reply apps/acqdat_api/lib/acqdat_api/dashboard_management/dashboard.ex:258]: We are using the datetime in inserted_at and updated_at fields.

Comment (BandanaPandey) [reply apps/acqdat_api/lib/acqdat_api/dashboard_management/dashboard.ex:258]: ok

**Pull request merged**